download.py

```diff
@@ -92,50 +134,70 @@ def get_zone_links(czds_base_url):
 
-    status_code = download_zone_response.status_code
+    try:
+        download_zone_response = do_get(url, access_token)
 
-    if status_code == 200:
-        # Try to get the filename from the header
-        _,option = cgi.parse_header(download_zone_response.headers['content-disposition'])
-        filename = option['filename']
+        status_code = download_zone_response.status_code
 
-        # If could get a filename from the header, then makeup one like [tld].txt.gz
-        if not filename:
-            filename = url.rsplit('/', 1)[-1].rsplit('.')[-2] + '.txt.gz'
+        if status_code == 200:
+            # Try to get the filename from the header
+            _,option = cgi.parse_header(download_zone_response.headers['content-disposition'])
+            filename = option['filename']
 
-        # This is where the zone file will be saved
-        path = '{0}/{1}'.format(output_directory, filename)
+            # If could get a filename from the header, then makeup one like [tld].txt.gz
+            if not filename:
+                filename = url.rsplit('/', 1)[-1].rsplit('.')[-2] + '.txt.gz'
 
-        with open(path, 'wb') as f:
-            for chunk in download_zone_response.iter_content(1024):
-                f.write(chunk)
+            # This is where the zone file will be saved
+            path = '{0}/{1}'.format(output_directory, filename)
 
-        print("{0}: Completed downloading zone to file {1}".format(str(datetime.datetime.now()), path))
+            with open(path, 'wb') as f:
+                for chunk in download_zone_response.iter_content(1024):
+                    f.write(chunk)
 
-    elif status_code == 401:
-        print("The access_token has been expired. Re-authenticate user {0}".format(username))
-        access_token = authenticate(username, password, authen_base_url)
-        download_one_zone(url, output_directory)
-    elif status_code == 404:
-        print("No zone file found for {0}".format(url))
-    else:
-        sys.stderr.write('Failed to download zone from {0} with code {1}\n'.format(url, status_code))
+            print("{0}: Completed downloading zone to file {1}".format(str(datetime.datetime.now()), path))
+
+        elif status_code == 401:
+            print("The access_token has been expired. Re-authenticate user {0}".format(username))
+            access_token = authenticate(username, password, authentication_base_url)
+            download_one_zone(url, output_directory)
+        elif status_code == 404:
+            print("No zone file found for {0}".format(url))
+        else:
+            sys.stderr.write('Failed to download zone from {0} with code {1}\n'.format(url, status_code))
+
+        # Make sure we throw an exception we can catch back down the line.
+        if status_code >= 400:
+            raise RuntimeError("Download failed.")
+
+    except:
+        # Don't just die here. We need to know that we failed.
+        raise
```

Comment:
Would it be at all useful to add the status code to the raise here?

Reply:
This reports the exception back to the loop that's responsible for downloading the zones and reporting status back to CloudWatch instead of just crashing here. That way we can post a 0 to CloudWatch for the file that's breaking.